File: sdk-platform/src/main/java/com/hezaro/wall/sdk/platform/player/LocalMediaPlayer.java
package com.hezaro.wall.sdk.platform.player;

import android.annotation.TargetApi;
import android.content.Context;
import android.net.Uri;
import android.os.Build;
import android.widget.Toast;
import com.google.android.exoplayer2.DefaultLoadControl;
import com.google.android.exoplayer2.DefaultRenderersFactory;
import com.google.android.exoplayer2.ExoPlaybackException;
import com.google.android.exoplayer2.ExoPlayerFactory;
import com.google.android.exoplayer2.PlaybackParameters;
import com.google.android.exoplayer2.Player;
import com.google.android.exoplayer2.RenderersFactory;
import com.google.android.exoplayer2.SimpleExoPlayer;
import com.google.android.exoplayer2.source.ConcatenatingMediaSource;
import com.google.android.exoplayer2.source.ExtractorMediaSource;
import com.google.android.exoplayer2.source.MediaSource;
import com.google.android.exoplayer2.source.TrackGroupArray;
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector;
import com.google.android.exoplayer2.trackselection.TrackSelectionArray;
import com.google.android.exoplayer2.trackselection.TrackSelector;
import com.google.android.exoplayer2.upstream.DataSource;
import com.google.android.exoplayer2.upstream.DefaultHttpDataSource;
import com.google.android.exoplayer2.upstream.DefaultHttpDataSourceFactory;
import com.google.android.exoplayer2.upstream.cache.CacheDataSource;
import com.google.android.exoplayer2.upstream.cache.CacheDataSourceFactory;
import com.google.android.exoplayer2.upstream.cache.LeastRecentlyUsedCacheEvictor;
import com.google.android.exoplayer2.upstream.cache.SimpleCache;
import com.hezaro.wall.data.model.Episode;
import com.hezaro.wall.data.model.Playlist;
import java.io.File;
import java.lang.ref.WeakReference;
import org.jetbrains.annotations.NotNull;
import timber.log.Timber;

public class LocalMediaPlayer extends MediaPlayer implements Player.EventListener {

    private static final String TAG = LocalMediaPlayer.class.getSimpleName();

    private static final int MAX_CACHE_SIZE = 250_000_000;

    private static final String PODCAST_CACHE_DIR = "podcast-cache";

    private final Context context;

    private SimpleExoPlayer exoPlayer;

    private Episode episode;

    private boolean isStreaming;

    private int mediaPlayerState;

    private Playlist playlist;

    private SimpleCache cache;

    public LocalMediaPlayer(WeakReference<MediaPlayerListener> mediaPlayerListener, Context context) {
        super(mediaPlayerListener);
        this.context = context;
        mediaPlayerListener.get().setPlayer(this);
        if (exoPlayer == null) {
            TrackSelector trackSelector = new DefaultTrackSelector();
            DefaultLoadControl loadControl = new DefaultLoadControl();
            RenderersFactory renderersFactory = new DefaultRenderersFactory(this.context);
            exoPlayer = ExoPlayerFactory
                    .newSimpleInstance(this.context, renderersFactory, trackSelector, loadControl);
            exoPlayer.addListener(this);
            mediaPlayerState = MediaPlayerState.STATE_IDLE;
        }
    }

    @Override
    public Player getPlayer() {
        return exoPlayer;
    }

    @Override
    public long getCurrentPosition() {
        return exoPlayer.getCurrentPosition();
    }

    @Override
    public long getDuration() {
        return exoPlayer.getDuration();
    }

    @Override
    public boolean isStreaming() {
        return isStreaming;
    }

    @NotNull
    @Override
    public Episode getCurrentEpisode() {
        return episode;
    }

    @Override
    public void concatPlaylist(@NotNull Playlist playlist) {
        this.playlist = playlist;
        final ConcatenatingMediaSource concatenatingMediaSource = new ConcatenatingMediaSource();
        for (Episode episode : playlist.getItems()) {
            concatenatingMediaSource.addMediaSource(buildMediaSource(episode));
        }
        exoPlayer.prepare(concatenatingMediaSource, true, false);
        exoPlayer.setPlayWhenReady(false);
    }

    @Override
    public void onTracksChanged(final TrackGroupArray trackGroups, final TrackSelectionArray trackSelections) {
        episode = playlist.getItem(exoPlayer.getCurrentWindowIndex());
        getListenerReference().notifyEpisode(episode);
    }

    @Override
    public void selectTrack(final Episode episode) {
        this.episode = episode;
        if (playlist != null) {
            int currentIndex = playlist.getIndex(episode);
            if (currentIndex >= 0 && exoPlayer.getCurrentWindowIndex() != currentIndex) {
                exoPlayer.seekTo(currentIndex, 0);
                exoPlayer.setPlayWhenReady(true);
            }
        }

    }

    @Override
    public boolean isPlaying() {
        return exoPlayer.getPlayWhenReady();
    }

    @Override
    public void next() {
        exoPlayer.next();
    }

    @Override
    public void previous() {
        exoPlayer.previous();
    }

    @Override
    public void resumePlayback() {
        exoPlayer.setPlayWhenReady(true);
    }

    @Override
    public void pausePlayback() {
        exoPlayer.setPlayWhenReady(false);
    }

    @Override
    public void stopPlayback() {
        exoPlayer.stop();
        isStreaming = false;
    }

    @Override
    public void seekTo(long position) {
        exoPlayer.seekTo(position);
    }

    @Override
    public void onDestroy() {
        Timber.tag(TAG).d("Tearing down");
        exoPlayer.release();
        exoPlayer.removeListener(this);
    }

    @Override
    public void onLoadingChanged(boolean isLoading) {

    }

    @Override
    public void onPlayerStateChanged(boolean playWhenReady, int playbackState) {
        String playbackStateStr;

        switch (playbackState) {
            case Player.STATE_BUFFERING:
                mediaPlayerState = MediaPlayerState.STATE_CONNECTING;
                playbackStateStr = "Buffering";
                break;
            case Player.STATE_ENDED:
                mediaPlayerState = MediaPlayerState.STATE_ENDED;
                playbackStateStr = "Ended";
                break;
            case Player.STATE_IDLE:
                mediaPlayerState = MediaPlayerState.STATE_IDLE;
                playbackStateStr = "Idle";
                break;
            case Player.STATE_READY:
                mediaPlayerState = playWhenReady ? MediaPlayerState.STATE_PLAYING :
                        MediaPlayerState.STATE_PAUSED;
                playbackStateStr = "Ready";
                break;
            default:
                mediaPlayerState = MediaPlayerState.STATE_IDLE;
                playbackStateStr = "Unknown";
                break;
        }
        getListenerReference().onStateChanged(mediaPlayerState);
        Timber.tag(TAG).d(String.format("ExoPlayer state changed: %s, Play When Ready: %s",
                playbackStateStr,
                String.valueOf(playWhenReady)));
    }

    @Override
    public void onPlayerError(ExoPlaybackException error) {
        Toast.makeText(context, error.getCause().getMessage(), Toast.LENGTH_SHORT).show();
        Timber.w(error, "Player error encountered");
        stopPlayback();

    }

    @TargetApi(Build.VERSION_CODES.M)
    @Override
    public void setPlaybackSpeed(float speed) {
        PlaybackParameters playbackParams = new PlaybackParameters(speed);
        exoPlayer.setPlaybackParameters(playbackParams);
    }

    private MediaSource buildMediaSource(Episode mEpisode) {
        final Uri uri = Uri.parse(mEpisode.getRemoteMediaUrl());
        final CacheDataSourceFactory dataSourceFactory = getCacheDataSource(
                new File(context.getCacheDir(), PODCAST_CACHE_DIR));
        isStreaming = true;

        if (uri != null) {
            Timber.tag(TAG).d("Playing from URI %s", uri);
            return new ExtractorMediaSource.Factory(dataSourceFactory).setTag(mEpisode.getDescription())
                    .createMediaSource(uri);
        }
        throw new IllegalStateException("Unable to build media source");
    }


    private CacheDataSourceFactory getCacheDataSource(File cacheDir) {
        if (cache == null) {
            cache = new SimpleCache(cacheDir, new LeastRecentlyUsedCacheEvictor(MAX_CACHE_SIZE));
        }

        DataSource.Factory upstream = new DefaultHttpDataSourceFactory("wall.userAgent", null,
                DefaultHttpDataSource.DEFAULT_CONNECT_TIMEOUT_MILLIS,
                DefaultHttpDataSource.DEFAULT_READ_TIMEOUT_MILLIS, true);
        return new CacheDataSourceFactory(cache, upstream,
                CacheDataSource.FLAG_IGNORE_CACHE_FOR_UNSET_LENGTH_REQUESTS,
                CacheDataSource.DEFAULT_MAX_CACHE_FILE_SIZE);
    }

}

File: sdk-platform/src/main/java/com/hezaro/wall/sdk/platform/player/LocalMediaPlayer.kt
package com.hezaro.wall.sdk.platform.player

import android.annotation.TargetApi
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.widget.Toast
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.crashlytics.android.Crashlytics
import com.google.android.exoplayer2.C
import com.google.android.exoplayer2.DefaultLoadControl
import com.google.android.exoplayer2.DefaultRenderersFactory
import com.google.android.exoplayer2.ExoPlaybackException
import com.google.android.exoplayer2.ExoPlayerFactory
import com.google.android.exoplayer2.PlaybackParameters
import com.google.android.exoplayer2.Player
import com.google.android.exoplayer2.SimpleExoPlayer
import com.google.android.exoplayer2.audio.AudioAttributes.Builder
import com.google.android.exoplayer2.source.ConcatenatingMediaSource
import com.google.android.exoplayer2.source.ExtractorMediaSource
import com.google.android.exoplayer2.source.MediaSource
import com.google.android.exoplayer2.source.TrackGroupArray
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector
import com.google.android.exoplayer2.trackselection.TrackSelectionArray
import com.hezaro.wall.data.model.Episode
import com.hezaro.wall.data.model.Playlist
import com.hezaro.wall.sdk.platform.player.download.PlayerDownloadHelper
import com.hezaro.wall.sdk.platform.utils.ACTION_EPISODE
import com.hezaro.wall.sdk.platform.utils.ACTION_EPISODE_GET
import com.hezaro.wall.sdk.platform.utils.ACTION_PLAYER
import com.hezaro.wall.sdk.platform.utils.ACTION_PLAYER_STATUS
import timber.log.Timber

class LocalMediaPlayer(private val context: Context) : MediaPlayer, Player.EventListener {

    private var exoPlayer: SimpleExoPlayer? = null

    private var episode: Episode? = null

    private var errorOccurred = false

    override var isStreaming: Boolean = false
        private set

    private var mediaPlayerState: Int = 0

    private var playlist: Playlist? = null

    private val concatenatingMediaSource = ConcatenatingMediaSource()

    override val player: Player
        get() = exoPlayer!!

    override val currentPosition: Long
        get() = exoPlayer!!.currentPosition

    override val duration: Long
        get() = exoPlayer!!.duration

    override val isPlaying: Boolean
        get() = exoPlayer!!.playWhenReady

    private val dataSourceFactory by lazy {
        PlayerDownloadHelper(context)
            .buildDataSourceFactory()
    }

    init {
        if (exoPlayer == null) {
            init()
        }
    }

    private lateinit var instanceListener: InstanceListener
    override fun setInstanceListener(instanceListener: InstanceListener) {
        this.instanceListener = instanceListener
    }

    override fun init() {
        val trackSelector = DefaultTrackSelector()
        val loadControl = DefaultLoadControl()
        val factory = DefaultRenderersFactory(this.context)
        exoPlayer = ExoPlayerFactory
            .newSimpleInstance(this.context, factory, trackSelector, loadControl)
        exoPlayer!!.addListener(this)
        val audioAttributes = Builder()
            .setUsage(C.USAGE_MEDIA)
            .setContentType(C.CONTENT_TYPE_MOVIE)
            .build()
        exoPlayer!!.setAudioAttributes(audioAttributes, true)
        mediaPlayerState = MediaPlayerState.STATE_IDLE
    }

    override fun getCurrentEpisode(): Episode? = episode

    override fun clearPlaylist() {
        playlist?.getItems()?.clear()
        concatenatingMediaSource.clear()
        playlist = null
    }

    override fun concatPlaylist(p: ArrayList<Episode>) {
        if (this.playlist == null) {
            this.playlist = Playlist(p)
            for (i in 0 until this.playlist!!.getItems().size) {
                concatenatingMediaSource.addMediaSource(buildMediaSource(playlist!!.getItem(i)))
            }
        } else {
            for (i in 0 until playlist!!.getItems().size) {
                val it = playlist!!.getItem(i)
                if (this.playlist!!.getIndex(it) == -1) {
                    this.playlist!!.addItem(it)
                    concatenatingMediaSource.addMediaSource(buildMediaSource(it))
                }
            }
        }
        val beReset = exoPlayer!!.currentWindowIndex > 0 || isPlaying
        if (!beReset) {
            if (!errorOccurred)
                exoPlayer!!.prepare(concatenatingMediaSource)
            else {
                exoPlayer!!.prepare(concatenatingMediaSource, false, false)
                if (::instanceListener.isInitialized) {
                    instanceListener.onNewInstance(exoPlayer, errorOccurred)
                }
                errorOccurred = false
            }
        }
    }

    override fun playPlaylist(p: ArrayList<Episode>, episode: Episode, readyToPlay: Boolean) {
        this.playlist?.getItems()?.clear()
        concatenatingMediaSource.clear()
        this.playlist = Playlist(p)
        for (i in 0 until this.playlist!!.getItems().size)
            concatenatingMediaSource.addMediaSource(buildMediaSource(playlist!!.getItem(i)))

        exoPlayer!!.prepare(concatenatingMediaSource)

        this.episode = episode
        broadcastEpisode(episode)
        val currentIndex = playlist!!.getIndex(episode)
        Timber.tag("MediaPlayer").i("Selected track index = $currentIndex")
        Timber.tag("MediaPlayer").i("episode.state (SEEK) = ${episode.state}")
        exoPlayer!!.seekTo(currentIndex, if (episode.state >= 0) episode.state else 0)

        exoPlayer!!.playWhenReady = readyToPlay
    }

    override fun playTrack(e: Episode) {
        playlist = Playlist(arrayListOf(e))
        concatenatingMediaSource.clear()
        concatenatingMediaSource.addMediaSource(buildMediaSource(e))
        exoPlayer!!.prepare(concatenatingMediaSource)
        this.episode = e
        broadcastEpisode(e)
        Timber.tag("MediaPlayer").i("Selected track index = ${playlist?.getIndex(episode!!)}")
        Timber.tag("MediaPlayer").i("episode.state (SEEK) = ${episode!!.state}")
        exoPlayer!!.seekTo(0, if (e.state >= 0) e.state else 0)
        exoPlayer!!.playWhenReady = true
    }

    override fun onTracksChanged(trackGroups: TrackGroupArray?, trackSelections: TrackSelectionArray?) {
        if (trackGroups!!.isEmpty && playlist!!.getItemCount() > 0) {
            episode = playlist!!.getItem(exoPlayer!!.currentWindowIndex)
            broadcastEpisode(episode!!)
        }
    }

    override fun selectTrack(episode: Episode) {
        this.episode = episode
        val currentIndex = playlist?.getIndex(episode) ?: -1
        if (playlist != null && currentIndex > 0) {
            if (exoPlayer!!.currentTimeline.windowCount > 0 &&
                exoPlayer!!.currentTimeline.windowCount >= currentIndex && exoPlayer!!.currentWindowIndex != currentIndex || currentIndex == 0
            ) {
                Timber.tag("MediaPlayer").i("Selected track index = $currentIndex")
                Timber.tag("MediaPlayer").i("episode.state (SEEK) = ${episode.state}")
                exoPlayer!!.seekTo(currentIndex, if (episode.state >= 0) episode.state else 0)
                exoPlayer!!.playWhenReady = true
            }
        } else playPlaylist(arrayListOf(episode), episode, false)
    }

    override fun next() {
        exoPlayer!!.next()
    }

    override fun previous() {
        exoPlayer!!.previous()
    }

    override fun resumePlayback() {
        exoPlayer!!.playWhenReady = true
    }

    override fun pausePlayback() {
        exoPlayer!!.playWhenReady = false
    }

    override fun stopPlayback() {
        exoPlayer!!.stop()
        isStreaming = false
    }

    override fun seekTo(position: Long) = exoPlayer!!.seekTo(position)

    override fun onLoadingChanged(isLoading: Boolean) {
    }

    override fun onPlayerStateChanged(playWhenReady: Boolean, playbackState: Int) {
        val playbackStateStr: String

        when (playbackState) {
            Player.STATE_BUFFERING -> {
                mediaPlayerState = MediaPlayerState.STATE_CONNECTING
                playbackStateStr = "Buffering"
            }
            Player.STATE_ENDED -> {
                mediaPlayerState = MediaPlayerState.STATE_ENDED
                playbackStateStr = "Ended"
            }
            Player.STATE_IDLE -> {
                mediaPlayerState = MediaPlayerState.STATE_IDLE
                playbackStateStr = "Idle"
            }
            Player.STATE_READY -> {
                mediaPlayerState = if (playWhenReady)
                    MediaPlayerState.STATE_PLAYING
                else
                    MediaPlayerState.STATE_PAUSED
                playbackStateStr = "Ready"
            }
            else -> {
                mediaPlayerState = MediaPlayerState.STATE_IDLE
                playbackStateStr = "Unknown"
            }
        }

        broadcastStatus(mediaPlayerState)

        Timber.tag(TAG).d(
            String.format(
                "ExoPlayer state changed: %s, Play When Ready: %s",
                playbackStateStr,
                playWhenReady.toString()
            )
        )
    }

    private fun broadcastEpisode(episode: Episode) {
        val intent = Intent(ACTION_EPISODE)
        intent.putExtra(ACTION_EPISODE_GET, episode)
        LocalBroadcastManager.getInstance(context).sendBroadcast(intent)
    }

    private fun broadcastStatus(status: Int) {
        val intent = Intent(ACTION_PLAYER)
        intent.putExtra(ACTION_PLAYER_STATUS, status)
        LocalBroadcastManager.getInstance(context).sendBroadcast(intent)
    }

    @TargetApi(Build.VERSION_CODES.M)
    override fun setPlaybackSpeed(speed: Float) {
        val playbackParams = PlaybackParameters(speed)
        exoPlayer!!.playbackParameters = playbackParams
    }

    private fun buildMediaSource(mEpisode: Episode): MediaSource {
        val uri = Uri.parse(mEpisode.source)
        isStreaming = true

        if (uri != null) {
            return ExtractorMediaSource.Factory(dataSourceFactory).setTag(mEpisode.description)
                .createMediaSource(uri)
        }
        throw IllegalStateException("Unable to build media source")
    }

    override fun onPlayerError(error: ExoPlaybackException) {
        errorOccurred = true
        var errorString = ""
        Crashlytics.logException(error)
        Crashlytics.setInt("PlayerStatus", exoPlayer!!.playbackState)
        Crashlytics.setBool("EpisodeIsNull", episode == null)
        Crashlytics.setBool("PlaylistIsNull", playlist == null)
        Crashlytics.setInt("PlaylistSize", if (playlist != null) playlist!!.getItemCount() else -1)

        when (error.type) {
            ExoPlaybackException.TYPE_RENDERER -> {
                stopPlayback()
                onDestroy()
                init()
                concatPlaylist(playlist!!.getItems())
                Timber.w(error, "Player error encountered -> TYPE_RENDERER ${error.rendererException.message}")
            }
            ExoPlaybackException.TYPE_SOURCE -> {
                val message = error.sourceException.message!!
                val errorCode = "[3-5][0-3][0-9]\\b".toRegex().find(message)?.value

                if (!errorCode.isNullOrEmpty()) {
                    stopPlayback()
                    onDestroy()
                    init()
                    concatPlaylist(playlist!!.getItems())
                    errorString = "امکان پخش این اپیزود وجود ندارد"
                } else if (message.contains("Unable to connect"))
                    errorString = "اینترنت دردسترس نیست"
            }
            ExoPlaybackException.TYPE_UNEXPECTED -> {
                stopPlayback()
                onDestroy()
                init()
                concatPlaylist(playlist!!.getItems())
                Timber.w(error, "Player error encountered -> TYPE_UNEXPECTED ${error.unexpectedException.message}")
            }
        }
        if (errorString.isNotEmpty())
            Toast.makeText(context, errorString, Toast.LENGTH_LONG).show()
    }

    override fun onDestroy() {
        exoPlayer!!.release()
        exoPlayer!!.removeListener(this)
        episode = null
    }

    companion object {

        private val TAG = LocalMediaPlayer::class.java.simpleName
    }
}
